src/tabs/Pixnflix/index.tsx

```diff
@@ -1,57 +1,252 @@
+
+type State = {
+  width: number;
+  height: number;
+  FPS: number;
+  mode: SideContentVideoDisplayMode;
+};
+
+type PixNFlix = {
+  toReplString: () => string;
+  init: (
+    video: HTMLVideoElement | null,
+    canvas: HTMLCanvasElement | null,
+    errorLogger: ErrorLogger
+  ) => void;
+  deinit: () => void;
+  startVideo: () => void;
+  snapPicture: () => void;
+  updateFPS: (fps: number) => void;
+  updateDimensions: (width: number, height: number) => void;
+};
 
 class Repeat extends React.Component<Props, State> {
   private $video: HTMLVideoElement | null = null;
 
   private $canvas: HTMLCanvasElement | null = null;
 
+  private pixNFlix: PixNFlix;
+
   constructor(props: any) {
     super(props);
-    this.state = {};
+    this.state = {
+      width: DEFAULT_WIDTH,
+      height: DEFAULT_HEIGHT,
+      FPS: DEFAULT_FPS,
+      mode: 'video' as SideContentVideoDisplayMode,
+    };
+    const { debuggerContext } = this.props;
+    this.pixNFlix = debuggerContext.result.value;
   }
 
   public componentDidMount() {
+    this.setupVideoService();
+    window.addEventListener('beforeunload', this.pixNFlix.deinit);
+  }
+
+  public componentWillUnmount() {
+    this.closeVideo();
+    window.removeEventListener('beforeunload', this.pixNFlix.deinit);
+  }
+
+  public setupVideoService = () => {
     if (this.$video && this.$canvas) {
-      // eslint-disable-next-line react/destructuring-assignment
-      this.props.debuggerContext.result.value.init(
-        this.$video,
-        this.$canvas,
-        this.printError
-      );
+      const { debuggerContext } = this.props;
+      this.pixNFlix = debuggerContext.result.value;
+      this.pixNFlix.init(this.$video, this.$canvas, this.printError);
     }
-  }
+  };
+
+  public closeVideo = () => {
+    this.pixNFlix.deinit();
+  };
 
-  public printError = () => {};
+  public handleStartVideo = () => {
+    this.pixNFlix.startVideo();
+  };
+
+  public handleSnapPicture = () => {
+    this.pixNFlix.snapPicture();
+  };
+
+  public handleWidthChange = (n: number) => {
+    if (n > 0 && n <= 500) {
+      this.setState((prevState) => ({
+        width: n,
+        height: prevState.height,
+      }));
+      const { height } = this.state;
+      this.handleUpdateDimensions(n, height);
+    }
+  };
+
+  public handleHeightChange = (m: number) => {
+    if (m > 0 && m <= 500) {
+      this.setState((prevState) => ({
+        width: prevState.width,
+        height: m,
+      }));
+      const { width } = this.state;
+      this.handleUpdateDimensions(width, m);
+    }
+  };
+
+  public handleFPSChange = (m: number) => {
+    // these magic numbers are based off video library
+    if (m > 2 && m < 30) {
+      this.setState({
+        FPS: m,
+      });
+      this.pixNFlix.updateFPS(m);
+    }
+  };
+
+  public handleUpdateDimensions = (n: number, m: number) => {
+    this.pixNFlix.updateDimensions(n, m);
+  };
+
+  public printError: ErrorLogger = () => {};

```

Comment:
I think this is a function that is going to be very common in many module tabs in the future so we gotta implement this and add the function to debuggerContext somehow.

Reply:
Yeah, will add this in future PRs